### PR TITLE
feat (Add a check for props for all constructs)

### PR DESCRIPTION
**Pull request by biffgaut**
*Issue #, if available:* [176](https://github.com/awslabs/aws-solutions-constructs/issues/176)

*Description of changes:*
Currently every construct performs its own validation of inputs. This validation is usually consistent across constructs but is inconsistent across services. For instance, if you send an existing VPC and VPC Props to a construct it will throw an error, but if you send an existing Lambda function and Lambda function props to a construct then the construct will ignore the props silently.

We need to have a consistent way to check inputs that ensures consistency across constructs as well as across services. In the process we will settle on explicitly throwing an error when incompatible props are submitted (eg - props and an existing resource).

(this is internal issues T131 and T137)

Use Case
Proposed Solution
This will be accomplished with a single prop validation routine that is called at the beginning of every construct.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-fkVQbXRiQi6A
* Commit ID: aa28f33772aa92c86093f860606583d0f4642ce0
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b5b89311-a636-46ee-a0d2-a475f6fd9923%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import * as sqs from '@aws-cdk/aws-sqs';
+import * as mediastore from '@aws-cdk/aws-mediastore';
+import * as s3 from '@aws-cdk/aws-s3';
+import * as sns from '@aws-cdk/aws-sns';
+import * as glue from '@aws-cdk/aws-glue';
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+import * as secretsmanager from "@aws-cdk/aws-secretsmanager";
+
+export interface VerifiedProps {
+  dynamoTableProps?: dynamodb.TableProps,
```
Comment (hnishar) [inline source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts:26]: Mark all the variables `readonly` ?

File: source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  readonly secretProps?: secretsmanager.SecretProps;
+
+}
+
+export function CheckProps(propsObject: VerifiedProps | any) {
+  let errorMessages = '';
+  let errorFound = false;
+
+  if (propsObject.dynamoTableProps && propsObject.existingTableObj) {
+    errorMessages += 'Cannot specify an existing DDB table AND DDB table props\n';
```
Comment (hnishar) [inline source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts:68]: Make the error message more user friendly by saying "Either provide `existingTableObj` or `dynamoTableProps`, not both", similar feedback for the rest of error messages

File: source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  deployDeadLetterQueue?: boolean,
+  deadLetterQueueProps?: sqs.QueueProps,
+
+  existingMediaStoreContainerObj?: mediastore.CfnContainer;
+  mediaStoreContainerProps?: mediastore.CfnContainerProps;
+
+  existingBucketObj?: s3.Bucket,
+  bucketProps?: s3.BucketProps,
+
+  // topicsProps is an incorrect attribute used in event-rule-sns that
```
Comment (hnishar) [inline source/patterns/@aws-solutions-constructs/core/lib/input-validation.ts:46]: Create an issue to track this change and make the update before we flip the switch for this pattern to `stable` ?

**Review by hnishar — CHANGES_REQUESTED** [review]

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-fkVQbXRiQi6A
* Commit ID: 4927f5767f123c5d4be2dbed14791e8b49323e04
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9d406469-1bd3-4a05-bdcb-58eae543b978%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-fkVQbXRiQi6A
* Commit ID: 27f72336dd4f046afc060d60b721440e6ae50c3b
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=07eed424-d9d3-4913-8657-4bb5f48c8d45%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-fkVQbXRiQi6A
* Commit ID: bd5c830b7a58a39fec3b8b53182c0f389ab39cba
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=89547f3b-2d10-4d9b-a202-4e1a9cb1b9f3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by biffgaut — COMMENTED** [review]
A few things to update, mostly naming.

**Review by hnishar — APPROVED** [review]

**Pull request merged**